lib/flame/code_sync.ex

```diff
@@ -0,0 +1,311 @@
+          basename = Path.basename(path),
+          [mod | _] = String.split(basename, ".beam"),
+          mod != basename,
+          do: Module.concat([mod])
+
+    %CodeSync{
+      current
+      | changed_paths: changed,
+        deleted_paths: deleted_paths,
+        purge_modules: module_to_purge,
+        apps_to_start: []
+    }
+  end
+
+  def package_to_stream(%CodeSync{} = code) do
+    verbose =
+      if code.verbose do
+        if !Enum.empty?(code.changed_paths),
+          do: log_verbose("packaging changed_paths: #{inspect(code.changed_paths)}")
+
+        if !Enum.empty?(code.apps_to_start),
+          do: log_verbose("sending apps_to_start: #{inspect(code.apps_to_start)}")
+
+        [:verbose]
+      else
+        []
+      end
+
+    out_stream =
+      if code.changed_paths != [] do
+        out_path = Path.join([code.tmp_dir.(), "flame_parent_code_sync_#{code.id}.tar.gz"])
+        dirs = for path <- code.changed_paths, uniq: true, do: String.to_charlist(path)
+        {:ok, tar} = :erl_tar.open(out_path, [:write, :compressed])
+        for dir <- dirs, do: :erl_tar.add(tar, dir, trim_leading_slash(dir), verbose)
+        :ok = :erl_tar.close(tar)
+
+        File.stream!(out_path, [], 2048)
+      end
+
+    %PackagedStream{
+      id: code.id,
+      tmp_dir: code.tmp_dir,
+      extract_dir: code.extract_dir,
+      sync_beam_hashes: code.sync_beam_hashes,
+      changed_paths: code.changed_paths,
+      deleted_paths: code.deleted_paths,
+      purge_modules: code.purge_modules,
+      apps_to_start: code.apps_to_start,
+      stream: out_stream,
+      verbose: code.verbose
+    }
+  end
+
+  defp trim_leading_slash([?/ | path]), do: path
+  defp trim_leading_slash([_ | _] = path), do: path
+
+  def extract_packaged_stream(%PackagedStream{} = pkg) do
+    if pkg.stream do
+      verbose = if pkg.verbose, do: [:verbose], else: []
+      extract_dir = pkg.extract_dir.()
+      target_tmp_path = Path.join([pkg.tmp_dir.(), "flame_child_code_sync_#{pkg.id}.tar.gz"])
+      flame_stream = File.stream!(target_tmp_path)
+      # transfer the file
+      Enum.into(pkg.stream, flame_stream)
+
+      # extract tar
+      :ok = :erl_tar.extract(target_tmp_path, [{:cwd, extract_dir}, :compressed | verbose])
+
+      # add code paths
+      :ok = add_code_paths_from_tar(pkg, extract_dir)
+
+      File.rm!(target_tmp_path)
+
+      # purge any deleted modules
+      for mod <- pkg.purge_modules do
+        if pkg.verbose && !Enum.empty?(pkg.purge_modules),
+          do: log_verbose("purging #{inspect(pkg.purge_modules)}")
+
+        :code.purge(mod)
+        :code.delete(mod)
+      end
+
+      # delete any deleted code paths, and prune empty dirs
+      for del_path <- pkg.deleted_paths do
+        File.rm!(del_path)
+        ebin_dir = Path.dirname(del_path)
+
+        if File.ls!(ebin_dir) == [] do
+          if pkg.verbose, do: log_verbose("deleting path #{ebin_dir}")
+          File.rm_rf!(ebin_dir)
```

Comment:
I have also removed some bangs, because we probably don't care if it fails, especially inside the flame. Your call.

```suggestion
      File.rm(target_tmp_path)

      # purge any deleted modules
      for mod <- pkg.purge_modules do
        if pkg.verbose && !Enum.empty?(pkg.purge_modules),
          do: log_verbose("purging #{inspect(pkg.purge_modules)}")

        :code.purge(mod)
        :code.delete(mod)
      end

      # delete any deleted code paths, and prune empty dirs
      for del_path <- pkg.deleted_paths do
        File.rm(del_path)
        ebin_dir = Path.dirname(del_path)

        if File.ls!(ebin_dir) == [] do
          if pkg.verbose, do: log_verbose("deleting path #{ebin_dir}")
          File.rm_rf(ebin_dir)
```